File: opencv-matching.rb
require 'ropencv'
include OpenCV

img1 = cv::imread ARGV.shift, cv::IMREAD_GRAYSCALE
img2 = cv::imread ARGV.shift, cv::IMREAD_GRAYSCALE

kp1 = Vector.new cv::KeyPoint
kp2 = Vector.new cv::KeyPoint
des1 = cv::Mat.new(3, 4, cv::CV_64FC1)
des2 = cv::Mat.new(3, 4, cv::CV_64FC1)

#detector, norm = cv::SIFT.new, cv::NORM_L2
detector, norm = cv::ORB.create, cv::NORM_HAMMING

detector.detect_and_compute img1, cv::Mat.new, kp1, des1
detector.detect_and_compute img2, cv::Mat.new, kp2, des2

matcher = cv::BFMatcher::new norm, true
matches = Vector.new cv::DMatch
matcher.match des1, des2, matches

img_matches = cv::Mat.new 3, 4, cv::CV_64FC1
top_matches = Vector.new(*matches.to_a[0..10])
cv::draw_matches img1, kp1, img2, kp2, top_matches, img_matches
cv::imshow 'matches', img_matches
cv::wait_key 0

File: opencv-matching.py
import cv2
import matplotlib.pyplot as plt
import sys

file1 = sys.argv[1]
file2 = sys.argv[2]

img1 = cv2.imread(file1, cv2.IMREAD_GRAYSCALE)
img2 = cv2.imread(file2, cv2.IMREAD_GRAYSCALE)

# detector, norm = cv2.xfeatures2d.SIFT_create(), cv2.NORM_L2
detector, norm = cv2.ORB_create(), cv2.NORM_HAMMING

kp1, des1 = detector.detectAndCompute(img1, None)
kp2, des2 = detector.detectAndCompute(img2, None)

matcher = cv2.BFMatcher(norm, crossCheck=True)

matches = matcher.match(des1, des2)
matches = sorted(matches, key=lambda x:x.distance)

img3=cv2.drawMatches(img1, kp1, img2, kp2, matches[:10], None)
plt.imshow(img3), plt.show()
